sdk/src/test/java/com/google/cloud/dataflow/sdk/runners/TemplatingDataflowPipelineRunnerTest.java

```diff
@@ -0,0 +1,122 @@
+import com.google.cloud.dataflow.sdk.testing.TestDataflowPipelineOptions;
+import com.google.cloud.dataflow.sdk.util.NoopPathValidator;
+import com.google.cloud.dataflow.sdk.util.TestCredential;
+
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.junit.rules.TemporaryFolder;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+import java.io.File;
+
+/**
+ * Tests for TemplatingDataflowPipelineRunner.
+ */
+@RunWith(JUnit4.class)
+public class TemplatingDataflowPipelineRunnerTest {
+
+  @Rule
+  public ExpectedLogs expectedLogs = ExpectedLogs.none(TemplatingDataflowPipelineRunner.class);
+
+  @Rule
+  public ExpectedException expectedThrown = ExpectedException.none();
+
+  @Rule
+  public final TemporaryFolder tmpDir = new TemporaryFolder();
+
+  /**
+   * Creates a mocked {@link DataflowPipelineJob} with the given {@code projectId} and {@code jobId}
+   * .
+   *
+   * <p>
+   * The return value may be further mocked.
+   */
+  private DataflowPipelineJob createMockJob(String projectId, String jobId) throws Exception {
+    DataflowPipelineJob mockJob = mock(DataflowPipelineJob.class);
+    when(mockJob.getProjectId()).thenReturn(projectId);
+    when(mockJob.getJobId()).thenReturn(jobId);
+    return mockJob;
+  }
+
+  /**
+   * Returns a {@link TemplatingDataflowPipelineRunner} that will return the provided a job to
+   * return. Some {@link PipelineOptions} will be extracted from the job, such as the project ID.
+   */
+  private TemplatingDataflowPipelineRunner createMockRunner(
+      DataflowPipelineJob job, String filePath) throws Exception {
+    DataflowPipelineRunner mockRunner = mock(DataflowPipelineRunner.class);
+    TestDataflowPipelineOptions options =
+        PipelineOptionsFactory.as(TestDataflowPipelineOptions.class);
+    options.setProject(job.getProjectId());
+    options.setDataflowJobFile(filePath);
+    when(mockRunner.run(isA(Pipeline.class))).thenReturn(job);
+
+    return new TemplatingDataflowPipelineRunner(mockRunner, options);
+  }
+
+  /**
+   * Tests that the {@link TemplatingDataflowPipelineRunner} returns normally when a template is
+   * successfully written.
+   */
+  @Test
+  public void testLoggedCompletion() throws Exception {
+    File existingFile = tmpDir.newFile();
+    createMockRunner(createMockJob("testJobDone-projectId", "testJobDone-jobId"),
+        existingFile.getPath()).run(DirectPipeline.createForTest());
+    expectedLogs.verifyInfo("Template successfully created");
+  }
+
+  /**
+   * Tests that the {@link TemplatingDataflowPipelineRunner} throws the appropriate exception when
+   * an output file is not writable.
+   */
+  @Test
+  public void testLoggedErrorForFile() throws Exception {
+    // TODO: Determine why this isn't failing.
+    // expectedThrown.expect(IOException.class);
+    createMockRunner(createMockJob("testJobDone-projectId", "testJobDone-jobId"), "/bad/path").run(
```

Comment:
I'm not sure this is the problem, but I've seen weird behaviors when using this kind of pattern to create mocks. I'd suggest:

1. Using fields instead of methods to create mocks, by annotating them with `@Mock` and using `MockitoAnnotations.initMocks` in the `@Before` `setUp` method.
2. Setting the expectations as described.

This moves most of the initialization/setup code into the `@Before`.